### PR TITLE
Add P24 as an UPE payment method

### LINKED ISSUE #2651: New UPE provider: Przelewy24
Add support for Przelewy24 (abbr: p24) as a WCPay UPE payment method.

https://stripe.com/docs/payments/p24

### Introduction of Przelewy24 from Stripe Docs: 

Przelewy24 is a Poland-based payment method aggregator that allows customers to complete transactions online using bank transfers and other methods. Bank transfers account for 80-90% of online payments in Poland and Przelewy24 provides a way for customers to pay with over 165 banks.

### PR DESCRIPTION
**Pull request by tpaksu**
Fixes #2651

#### Changes proposed in this Pull Request

TBD

Add Przelewy24 (P24) as an UPE payment method to WooCommerce Payments.

This PR adds the P24 payment method as an UPE provider to WooCommerce Payments. P24 is the most popular payment method in Poland, it supports PLN and EUR as payment currencies, also from their website, they support:

- card transaction processing on the merchant's website, the option of adapting the card form to the website/store layout,
- One Click Payment - payment with a single click - with no need for the buyer to create an account in the payment service provider's system,
- supporting transactions in 50 currencies with no currency conversion fees - 1 to 1 payment,
- DCC (Dynamic Currency Conversion) - this solution allows customers to make transactions in foreign currencies. A foreign card-holder may choose the currency for a given transaction but is charged the amount in his native currency. When this type of payment is made, the merchant receives a bonus - a refund of up to 1% of the transaction value,
- recurring payments which guarantee timely settlements with merchants who offer subscriptions, passes, etc.,
- card payments over the phone - IVR
- transfer of risk in case of fraud.

After the implementation, these screens now contain P24 elements:

##### Add Payment Method Modal
Add Payment Methods

##### Enabled Payment Methods List
Enabled Payment Methods

##### Payment Screen
Payment Screen

##### Transactions List
Transactions List

##### Transaction Details
Transaction Details

##### Refunds List
Refund

##### Refund Details
Refund Details

#### Testing instructions

* Go to **WooCommerce > Settings > Payments** And on this screen, find **Payment Methods** metabox (ensure you have grouped settings enabled)
* Check if you can see the P24 icon on the right of the **Add Payment Method** button.
* Click the **Add Payment Method** button, select **Przelewy24 (P24)** and click **Add selected**.
* Check if **Przelewy24 (P24)** is added to the **Payment Methods** list.
* Go to your store and select your customer currency to EUR. Or if you don't have customer multi currency, set your store currency to EUR from **WooCommerce > Settings > General** page.
* Add some items to the cart, and proceed to checkout. 
* On the checkout page, select **Use a different payment method** radio button, and wait for the UPE iframe to render.
* Check if you can see Przelewy24 (P24) on the available payment methods list in the iframe.
* Select P24 and on the new form below, select a bank.
* Click **Place Order** button to complete the payment.
* This should redirect you to Stripe's P24 test transaction page. 

image

* On that page, you have two options:
    - Click **Fail test payment** and check if a Payment error is displayed after that on the checkout page.
    - Click **Authorize test payment** and check if you get redirected to the order received page. 

* After you completed the payment, go to admin page **Payments > Transactions** (you may need to sync the transactions via CLI, or have server hook listener open to update it) and check if you can see the transaction with the P24 icon on it. 
* Click that row and check the details if the details are correct.

**Note:** Whatever you select as a bank on P24 payment page, it returns a hardcoded response as the success message with this data:

```json
"payment_method_details": {
    "p24": {
        "bank": "ing",
        "reference": "P24-N01-101-101 STRIPE POLAND",
        "verified_name": "Jenny Rosen"
    },
    "type": "p24"
},
```

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

### CODE REVIEW

**Comment by tpaksu** [comment]
The calculations on the refunded transactions detail seem a little bit off? Don't you think? While net should be 0, it shows $-2.79 USD.

**Review by jessepearson — CHANGES_REQUESTED** [review]
It looks like some legacy files and changes are still in here, I noted where those were.

There was a test that wasn't updated:
client/additional-methods-setup/methods-selector/test/add-payment-methods-task.test.js

There was stuff I missed in mine that I went back and fixed after seeing yours 😀 

I think once this has develop merged into it that some of the tests will fail due to they look for specific counts to calls and array indexes.

File: includes/payment-methods/class-p24-payment-method.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	 * Constructor for P24 payment method
+	 *
+	 * @param WC_Payments_Token_Service $token_service Token class instance.
+	 */
+	public function __construct( $token_service ) {
+		parent::__construct( $token_service );
+		$this->stripe_id   = 'p24';
+		$this->title       = 'p24';
+		$this->is_reusable = false;
+		$this->currencies  = [ 'EUR', 'PLN' ];
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply includes/payment-methods/class-p24-payment-method.php:27]: Yes it also shows on the PLN currency. I'll check those files.

[the remark this answers is not in the document]
Comment (tpaksu) [reply includes/payment-methods/class-p24-payment-method.php:27]: I've changed the code checking for additional currencies to support multiple currencies in both JavaScript and PHP.

**Comment by jessepearson** [comment]
> The calculations on the refunded transactions detail seem a little bit off? Don't you think? While net should be 0, it shows $-2.79 USD.

Fees do not get refunded, the merchant is still charged for the transaction.

**Comment by tpaksu** [comment]
> > The calculations on the refunded transactions detail seem a little bit off? Don't you think? While net should be 0, it shows $-2.79 USD.
> 
> Fees do not get refunded, the merchant is still charged for the transaction.

Oh, I didn't know that. Thanks for the clarification. Then everything's normal I guess.

**Review by jessepearson — CHANGES_REQUESTED** [review]
Looking good! A few more changes and I think this will be ready to merge.

File: client/additional-methods-setup/upe-preview-methods-selector/test/add-payment-methods-task.test.js [hunk 2]
```diff
@@ -101,14 +102,17 @@ describe( 'AddPaymentMethodsTask', () => {
 		);
 
 		expect(
-			screen.queryByText( /we\'ll add Euro \(€\) to your store/ )
+			screen.queryByText( /we\'ll add EUR to your store/ )
```
[the remark this answers is not in the document]
Comment (tpaksu) [reply client/additional-methods-setup/upe-preview-methods-selector/test/add-payment-methods-task.test.js:105]: Do we have a sigma for this, or is this something we decided? I guess we don't have the symbol in the enabled currencies array on JS side.

[the remark this answers is not in the document]
Comment (jessepearson) [reply client/additional-methods-setup/upe-preview-methods-selector/test/add-payment-methods-task.test.js:105]: > Do we have a Figma for this, or is this something we decided?
I can't find anything where it was decided by design, so I think it was something decided on the dev side. Either way, we should get input from design on it and open another issue if needed.

**Comment by tpaksu** [comment]
Now the currencies and providers are merged into a single notice box (4901c11):

image

File: client/additional-methods-setup/upe-preview-methods-selector/test/add-payment-methods-task.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (tpaksu) [reply client/additional-methods-setup/upe-preview-methods-selector/test/add-payment-methods-task.test.js]: Ok I've fixed that.

**Comment by jessepearson** [comment]
Noting here that I created https://github.com/Automattic/woocommerce-payments/issues/2720 to address the copy phrasing for the notice regarding adding currencies.

**Review by jessepearson — APPROVED** [review]
Able to place orders and perform refunds.
Transaction data shows correct in admin.
Currencies get added as they should.
Good to go 👍  :shipit:

**Pull request merged**